// @midwayjs/orm/dist/repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCustomRepository = exports.getMongoRepository = exports.getTreeRepository = exports.getRepository = void 0;
const core_1 = require("@midwayjs/core");
const _1 = require(".");
function getRepository(context, args) {
    return clzz => {
        const getConnection = context.get(_1.CONNECTION_KEY);
        return getConnection().getRepository(clzz);
    };
}
exports.getRepository = getRepository;
function getTreeRepository(context, args) {
    return clzz => {
        const getConnection = context.get(_1.CONNECTION_KEY);
        return getConnection().getTreeRepository(clzz);
    };
}
exports.getTreeRepository = getTreeRepository;
function getMongoRepository(context, args) {
    return clzz => {
        const getConnection = context.get(_1.CONNECTION_KEY);
        return getConnection().getMongoRepository(clzz);
    };
}
exports.getMongoRepository = getMongoRepository;
function getCustomRepository(context, args) {
    return clzz => {
        const getConnection = context.get(_1.CONNECTION_KEY);
        return getConnection().getCustomRepository(clzz);
    };
}
exports.getCustomRepository = getCustomRepository;
(0, core_1.providerWrapper)([
    {
        id: 'getRepository',
        provider: getRepository,
    },
    {
        id: 'getTreeRepository',
        provider: getTreeRepository,
    },
    {
        id: 'getMongoRepository',
        provider: getMongoRepository,
    },
    {
        id: 'getCustomRepository',
        provider: getCustomRepository,
    },
]);
